File: sample-conference-videochat/sample-conference-videochat/Components/Users/ViewControllers/MainTableViewController/MainTableViewController.m
//
//  MainTableViewController.m
//  sample-multiconference-videochat
//
//  Copyright (c) 2017 QuickBlox. All rights reserved.
//

#import "MainTableViewController.h"
#import <Quickblox/Quickblox.h>
#import "QBCore.h"
#import "SessionSettingsViewController.h"
#import "SVProgressHUD.h"
#import "DialogsDataSource.h"
#import "UsersDataSource.h"
#import "QBDataFetcher.h"
#import "UsersViewController.h"
#import "CallViewController.h"
#import "QBAVCallPermissions.h"

typedef NS_ENUM(NSUInteger, CallSenderValue) {
    CallSenderValueDialogInstance,
    CallSenderValueConferenceType
};

static NSString * const kSettingsSegue = @"PresentSettingsViewController";
static NSString * const kUsersSegue = @"PresentUsersViewController";
static NSString * const kCallSegue = @"PresentCallViewController";
static NSString * const kSceneSegueAuth = @"SceneSegueAuth";

@interface MainTableViewController () <SettingsViewControllerDelegate, QBCoreDelegate, UsersViewControllerDelegate, DialogsDataSourceDelegate>

@property (strong, nonatomic) DialogsDataSource *dialogsDataSource;
@property (strong, nonatomic) UsersDataSource *usersDataSource;

@end

@implementation MainTableViewController

// MARK: Lifecycle

- (void)dealloc {
    
    ILog(@"%@ - %@",  NSStringFromSelector(_cmd), self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [Core addDelegate:self];
    
    // Reachability
    __weak __typeof(self)weakSelf = self;
    Core.networkStatusBlock = ^(QBNetworkStatus status) {
        if (status != QBNetworkStatusNotReachable) {
            [weakSelf fetchData];
        }
    };
    
    [self configureNavigationBar];
    [self configureTableViewController];
    [self fetchData];
    
    // adding refresh control task
    if (self.refreshControl) {
        
        [self.refreshControl addTarget:self
                                action:@selector(fetchData)
                      forControlEvents:UIControlEventValueChanged];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (self.refreshControl.refreshing) {
        [self.tableView setContentOffset:CGPointMake(0, -self.refreshControl.frame.size.height) animated:NO];
    }
}

// MARK: UI Configuration

- (void)configureTableViewController {
    
    self.dialogsDataSource = [DialogsDataSource dialogsDataSource];
    self.dialogsDataSource.delegate = self;
    self.usersDataSource = [UsersDataSource usersDataSource];
    self.tableView.dataSource = self.dialogsDataSource;
    self.tableView.rowHeight = 76;
    [self.refreshControl beginRefreshing];
}

- (void)configureNavigationBar {
    
    UIBarButtonItem *settingsButtonItem =
    [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"ic-settings"]
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(didPressSettingsButton:)];
    
    self.navigationItem.leftBarButtonItem = settingsButtonItem;
    
    UIBarButtonItem *usersButtonItem =
    [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"new-message"]
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(didPressUsersButton:)];
    
    self.navigationItem.rightBarButtonItem = usersButtonItem;
    
    //Custom label
    NSString *roomName = [NSString stringWithFormat:@"%@", Core.currentUser.tags.firstObject];
    NSString *userName = [NSString stringWithFormat:@"Logged in as %@", Core.currentUser.fullName];
    NSString *titleString = [NSString stringWithFormat:@"%@\n%@", roomName, userName];
    
    NSMutableAttributedString *attrString = [[NSMutableAttributedString alloc] initWithString:titleString];
    NSRange roomNameRange = [titleString rangeOfString:roomName];
    [attrString addAttribute:NSFontAttributeName
                       value:[UIFont boldSystemFontOfSize:16.0f]
                       range:roomNameRange];
    
    NSRange userNameRange = [titleString rangeOfString:userName];
    [attrString addAttribute:NSFontAttributeName
                       value:[UIFont systemFontOfSize:12.0f]
                       range:userNameRange];
    [attrString addAttribute:NSForegroundColorAttributeName
                       value:[UIColor grayColor]
                       range:userNameRange];
    
    UILabel *titleView = [[UILabel alloc] initWithFrame:CGRectZero];
    titleView.numberOfLines = 2;
    titleView.attributedText = attrString;
    titleView.textAlignment = NSTextAlignmentCenter;
    [titleView sizeToFit];
    
    self.navigationItem.titleView = titleView;
}

// MARK: Actions

- (BOOL)hasConnectivity {
    
    BOOL hasConnectivity = Core.networkStatus != QBNetworkStatusNotReachable;
    
    if (!hasConnectivity) {
        [self showAlertViewWithMessage:NSLocalizedString(@"Please check your Internet connection", nil)];
    }
    
    return hasConnectivity;
}

- (void)showAlertViewWithMessage:(NSString *)message {
    
    UIAlertController *alertController =
    [UIAlertController alertControllerWithTitle:nil
                                        message:message
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Ok", nil)
                                                        style:UIAlertActionStyleDefault
                                                      handler:nil]];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)didPressSettingsButton:(UIBarButtonItem *)item {
    
    [self performSegueWithIdentifier:kSettingsSegue sender:item];
}

- (void)didPressUsersButton:(UIBarButtonItem *)item {
    
    [self performSegueWithIdentifier:kUsersSegue sender:item];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([segue.identifier isEqualToString:kSettingsSegue]) {
        
        SessionSettingsViewController *settingsViewController =
        (id)((UINavigationController *)segue.destinationViewController).topViewController;
        settingsViewController.delegate = self;
    }
    else if ([segue.identifier isEqualToString:kUsersSegue]) {
        
        UsersViewController *usersViewController =
        (id)segue.destinationViewController;
        usersViewController.dataSource = self.usersDataSource;
        usersViewController.delegate = self;
    }
    else if ([segue.identifier isEqualToString:kCallSegue]) {
        
        CallViewController *callVC = (id)segue.destinationViewController;
        callVC.chatDialog = sender[CallSenderValueDialogInstance];
        callVC.conferenceType = [sender[CallSenderValueConferenceType] integerValue];
        callVC.usersDataSource = self.usersDataSource;
    }
}

// MARK: - DialogsDataSourceDelegate

- (void)dialogsDataSource:(DialogsDataSource *)dialogsDataSource dialogCellDidTapListener:(__kindof UITableViewCell *)dialogCell {
    
    [self joinDialogFromDialogCell:dialogCell conferenceType:0];
}

- (void)dialogsDataSource:(DialogsDataSource *)dialogsDataSource dialogCellDidTapAudio:(__kindof UITableViewCell *)dialogCell {
    
    [self joinDialogFromDialogCell:dialogCell conferenceType:QBRTCConferenceTypeAudio];
}

- (void)dialogsDataSource:(DialogsDataSource *)dialogsDataSource dialogCellDidTapVideo:(__kindof UITableViewCell *)dialogCell {
    
    [self joinDialogFromDialogCell:dialogCell conferenceType:QBRTCConferenceTypeVideo];
}

- (void)dialogsDataSource:(DialogsDataSource *)dialogsDataSource commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if ([self hasConnectivity]
        && editingStyle == UITableViewCellEditingStyleDelete) {
        
        [SVProgressHUD show];
        QBChatDialog *chatDialog = self.dialogsDataSource.objects[indexPath.row];
        __weak __typeof(self)weakSelf = self;
        [QBRequest deleteDialogsWithIDs:[NSSet setWithObject:chatDialog.ID] forAllUsers:NO successBlock:^(QBResponse * _Nonnull response, NSArray<NSString *> * _Nullable deletedObjectsIDs, NSArray<NSString *> * _Nullable notFoundObjectsIDs, NSArray<NSString *> * _Nullable wrongPermissionsObjectsIDs) {
            
            __typeof(weakSelf)strongSelf = weakSelf;
            NSMutableArray *dialogs = [strongSelf.dialogsDataSource.objects mutableCopy];
            [dialogs removeObject:chatDialog];
            strongSelf.dialogsDataSource.objects = [dialogs copy];
            [self.tableView reloadData];
            [SVProgressHUD dismiss];
            
        } errorBlock:^(QBResponse * _Nonnull response) {
            
            [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"%@", response.error.reasons]];
        }];
    }
}

// MARK: UsersViewControllerDelegate

- (void)usersViewController:(UsersViewController *)usersViewController didCreateChatDialog:(QBChatDialog *)chatDialog {
    
    NSMutableArray *mutableObjecs = [self.dialogsDataSource.objects mutableCopy];
    [mutableObjecs addObject:chatDialog];
    [self.dialogsDataSource setObjects:[mutableObjecs copy]];
    [self.tableView reloadData];
}

// MARK: QBCoreDelegate

- (void)coreDidLogout:(QBCore *)core {
    
    [SVProgressHUD dismiss];
    //Dismiss Settings view controller
    [self dismissViewControllerAnimated:NO completion:nil];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self performSegueWithIdentifier:kSceneSegueAuth sender:nil];
    });
}

- (void)core:(QBCore *)core error:(NSError *)error domain:(ErrorDomain)domain {
    
    if (domain == ErrorDomainLogOut) {
        [SVProgressHUD showErrorWithStatus:error.localizedDescription];
    }
}

// MARK: SettingsViewControllerDelegate

- (void)settingsViewController:(SessionSettingsViewController *)vc didPressLogout:(id)sender {
    
    [SVProgressHUD showWithStatus:NSLocalizedString(@"Logout...", nil)];
    [Core logout];
}

// MARK: Private

- (void)fetchData {
    
    __weak __typeof(self)weakSelf = self;
    dispatch_group_t dataGroup = dispatch_group_create();
    
    dispatch_group_enter(dataGroup);
    [QBDataFetcher fetchDialogs:^(NSArray *dialogs) {
        
        dispatch_group_leave(dataGroup);
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf.dialogsDataSource setObjects:dialogs];
        [strongSelf.tableView reloadData];
    }];
    
    dispatch_group_enter(dataGroup);
    [QBDataFetcher fetchUsers:^(NSArray *users) {
        
        dispatch_group_leave(dataGroup);
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf.usersDataSource setObjects:users];
    }];
    
    dispatch_group_notify(dataGroup, dispatch_get_main_queue(), ^{
        
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf.refreshControl endRefreshing];
    });
}

- (void)joinDialogFromDialogCell:(UITableViewCell *)cell conferenceType:(QBRTCConferenceType)conferenceType {
    if ([self hasConnectivity]) {
        
        if (conferenceType > 0) {
            [QBAVCallPermissions checkPermissionsWithConferenceType:conferenceType completion:^(BOOL granted) {
                
                if (granted) {
                    
                    NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
                    QBChatDialog *chatDialog = self.dialogsDataSource.objects[indexPath.row];
                    [self performSegueWithIdentifier:kCallSegue sender:@[chatDialog, @(conferenceType)]];
                }
            }];
        }
        else {
            NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
            QBChatDialog *chatDialog = self.dialogsDataSource.objects[indexPath.row];
            [self performSegueWithIdentifier:kCallSegue sender:@[chatDialog, @(conferenceType)]];
        }
    }
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/ViewControllers/MainTableViewController/MainTableViewController.swift
//
//  MainTableViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 11.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import SVProgressHUD
import Quickblox
import QuickbloxWebRTC

struct MainSegueConstant {
    static let settings = "PresentSettingsViewController"
    static let users = "PresentUsersViewController"
    static let call = "PresentCallViewController"
    static let sceneAuth = "SceneSegueAuth"
}

struct MainAlertConstant {
    static let checkInternet = NSLocalizedString("Please check your Internet connection", comment: "")
    static let okAction = NSLocalizedString("Ok", comment: "")
    static let logout = NSLocalizedString("Logout...", comment: "")
}

struct CallSettings {
    var conferenseType: QBRTCConferenceType?
    var chatDialog: QBChatDialog
}

class MainTableViewController: UITableViewController, SettingsViewControllerDelegate, QBCoreDelegate,
DialogsDataSourceDelegate, UsersViewControllerDelegate {
    
    //MARK: Variables
    let core = QBCore.instance
    private lazy var dialogsDataSource:DialogsDataSource = {
        let dialogsDataSource = DialogsDataSource()
        dialogsDataSource.delegate = self
        return dialogsDataSource
    }()
    var usersDataSource = UsersDataSource()
    
    // MARK: Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        core.addDelegate(self)
        
        // Reachability
        core.networkStatusBlock = { [weak self] status in
            if status != NetworkConnectionStatus.notConnection {
                self?.fetchData()
            }
        }
        
        configureNavigationBar()
        configureTableViewController()
        fetchData()
        
        // adding refresh control task
        if let refreshControl = self.refreshControl {
            refreshControl.addTarget(self, action: #selector(fetchData), for: .valueChanged)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let refreshControl = self.refreshControl, refreshControl.isRefreshing == true {
            let contentOffset = CGPoint(x: 0, y: -refreshControl.frame.size.height)
            tableView.setContentOffset(contentOffset, animated: false)
        }
    }
    
    deinit {
        debugPrint("deinit \(self)")
    }
    
    // MARK: UI Configuration
    func configureTableViewController() {
        tableView.dataSource = dialogsDataSource
        tableView.rowHeight = 76.0
        refreshControl?.beginRefreshing()
    }
    
    func configureNavigationBar() {
        let settingsButtonItem = UIBarButtonItem(image: UIImage(named: "ic-settings"),
                                                 style: .plain,
                                                 target: self,
                                                 action: #selector(didPressSettingsButton(_:)))
        navigationItem.leftBarButtonItem = settingsButtonItem
        
        let usersButtonItem = UIBarButtonItem(image: UIImage(named: "new-message"),
                                              style: .plain,
                                              target: self,
                                              action: #selector(didPressUsersButton(_:)))
        navigationItem.rightBarButtonItem = usersButtonItem
        
        //Custom label
        var userName = "Logged in as "
        var roomName = ""
        var titleString = ""
        if let currentUser = core.currentUser,
            let fullname = currentUser.fullName,
            let tags = currentUser.tags,
            tags.isEmpty == false,
            let name = tags.first {
            roomName = name
            userName = userName + fullname
            titleString = roomName + "\n" + userName
        }

        let attrString = NSMutableAttributedString(string: titleString)
        let roomNameRange: NSRange = (titleString as NSString).range(of: roomName )
        attrString.addAttribute(.font, value: UIFont.boldSystemFont(ofSize: 16.0), range: roomNameRange)
        
        let userNameRange: NSRange = (titleString as NSString).range(of: userName)
        attrString.addAttribute(.font, value: UIFont.systemFont(ofSize: 12.0), range: userNameRange)
        attrString.addAttribute(.foregroundColor, value: UIColor.gray, range: userNameRange)
        
        let titleView = UILabel(frame: CGRect.zero)
        titleView.numberOfLines = 2
        titleView.attributedText = attrString
        titleView.textAlignment = .center
        titleView.sizeToFit()
        
        navigationItem.titleView = titleView
    }
    
    // MARK: Actions
    func hasConnectivity() -> Bool {
        let status = core.networkConnectionStatus()
        guard status != NetworkConnectionStatus.notConnection else {
            showAlertView(withMessage: MainAlertConstant.checkInternet)
            return false
        }
        return true
    }
    
    func showAlertView(withMessage message: String?) {
        let alertController = UIAlertController(title: nil, message: message, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: MainAlertConstant.okAction, style: .default,
                                                handler: nil))
        present(alertController, animated: true)
    }
    
    @objc func didPressSettingsButton(_ item: UIBarButtonItem?) {
        performSegue(withIdentifier: MainSegueConstant.settings, sender: item)
    }
    
    @objc func didPressUsersButton(_ item: UIBarButtonItem?) {
        performSegue(withIdentifier: MainSegueConstant.users, sender: item)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        switch segue.identifier {
        case MainSegueConstant.settings:
            let settingsViewController = (segue.destination as? UINavigationController)?.topViewController
                as? SessionSettingsViewController
            settingsViewController?.delegate = self
            
        case MainSegueConstant.users:
            let usersViewController = segue.destination as? UsersViewController
            usersViewController?.dataSource = usersDataSource
            usersViewController?.delegate = self
            
        case MainSegueConstant.call:
            guard let settings = sender as? CallSettings else { return }
            let callViewController = segue.destination as? CallViewController
            callViewController?.chatDialog = settings.chatDialog
            callViewController?.conferenceType = settings.conferenseType
            callViewController?.usersDataSource = usersDataSource
            
        default:
            break
        }
    }
    
    //MARK: - DialogsDataSourceDelegate
    func dialogsDataSource(_ dialogsDataSource: DialogsDataSource?,
                           dialogCellDidTapListener dialogCell: UITableViewCell?) {
        joinDialog(fromDialogCell: dialogCell, conferenceType: nil)
    }
    
    func dialogsDataSource(_ dialogsDataSource: DialogsDataSource?,
                           dialogCellDidTapAudio dialogCell: UITableViewCell?) {
        joinDialog(fromDialogCell: dialogCell, conferenceType: QBRTCConferenceType.audio)
    }
    
    func dialogsDataSource(_ dialogsDataSource: DialogsDataSource?,
                           dialogCellDidTapVideo dialogCell: UITableViewCell?) {
        joinDialog(fromDialogCell: dialogCell, conferenceType: QBRTCConferenceType.video)
    }
    
    func dialogsDataSource(_ dialogsDataSource: DialogsDataSource?,
                           commit editingStyle: UITableViewCell.EditingStyle,
                           forRowAt indexPath: IndexPath?) {
        
        guard let indexPath = indexPath, let dataSource = dialogsDataSource else { return }
        let deleteDialog = dataSource.objects[indexPath.row]
        guard let deleteDialogId = deleteDialog.id else { return }
        if hasConnectivity() && editingStyle == .delete {
            SVProgressHUD.show()
            
            let chatDialogIDs: Set = [deleteDialogId]
            QBRequest.deleteDialogs(withIDs: chatDialogIDs,
                                    forAllUsers: false,
                                    successBlock: { [weak self] response,
                                        deletedObjectsIDs,
                                        notFoundObjectsIDs,
                                        wrongPermissionsObjectsIDs in
                                        
                                        guard let `self` = self else { return }
                                        //remove deleted dialog from datasource
                                        let dialogs = self.dialogsDataSource.objects
                                        let filteredDialogs = dialogs.filter({$0 != deleteDialog})
                                        self.dialogsDataSource.updateObjects(filteredDialogs)
                                        self.tableView.reloadData()
                                        SVProgressHUD.dismiss()
                }, errorBlock: { response in
                    SVProgressHUD.showError(withStatus: "\(String(describing: response.error?.reasons))")
            })
        }
    }
    
    // MARK: UsersViewControllerDelegate
    func usersViewController(_ usersViewController: UsersViewController?,
                             didCreateChatDialog chatDialog: QBChatDialog?) {
        guard let chatDialog = chatDialog else { return }
        dialogsDataSource.addObjects([chatDialog])
        tableView.reloadData()
    }
    
    // MARK: QBCoreDelegate
    func coreDidLogout(_ core: QBCore?) {
        SVProgressHUD.dismiss()
        //Dismiss Settings view controller
        dismiss(animated: false)
        DispatchQueue.main.async(execute: {
            self.performSegue(withIdentifier: MainSegueConstant.sceneAuth, sender: nil)
        })
    }
    
    func core(_ core: QBCore?, error: Error?, domain: ErrorDomain) {
        if domain == ErrorDomain.ErrorDomainLogOut {
            SVProgressHUD.showError(withStatus: error?.localizedDescription)
        }
    }
    
    // MARK: SettingsViewControllerDelegate
    func settingsViewController(_ vc: SessionSettingsViewController?, didPressLogout sender: Any?) {
        SVProgressHUD.show(withStatus: MainAlertConstant.logout)
        core.logout()
    }
    
    func coreDidLogin(_ core: QBCore) {
        debugPrint("coreDidLogin")
    }
    
    func coreDidLogout(_ core: QBCore) {
        debugPrint("coreDidLogin")
    }
    
    func core(_ core: QBCore, _ loginStatus: String) {
        debugPrint("coreDidLogin")
    }
    
    func core(_ core: QBCore, _ error: Error, _ domain: ErrorDomain) {
        debugPrint("coreDidLogin")
    }
    
    // MARK: Private
    @objc private func fetchData() {
        let dataGroup = DispatchGroup()
        dataGroup.enter()
        QBDataFetcher.fetchDialogs({ [weak self] dialogs in
            dataGroup.leave()
            if let dialogs = dialogs, dialogs.isEmpty == false {
                self?.dialogsDataSource.updateObjects(dialogs)
                self?.tableView.reloadData()
            }
        })
        dataGroup.enter()
        QBDataFetcher.fetchUsers({ [weak self] users in
            dataGroup.leave()
            guard let users = users else { return }
            self?.usersDataSource.updateObjects(users)
        })
        dataGroup.notify(queue: DispatchQueue.main) {
            self.refreshControl?.endRefreshing()
        }
    }
    
    private func joinDialog(fromDialogCell cell: UITableViewCell?, conferenceType: QBRTCConferenceType?) {
        guard hasConnectivity() == true,
            let cell = cell,
            let indexPath = self.tableView.indexPath(for: cell) else {
                return
        }
        let chatDialog = self.dialogsDataSource.objects[indexPath.row]
        let callSettings = CallSettings(conferenseType: conferenceType, chatDialog: chatDialog)
        guard let conferenceType = conferenceType else {
            // will join to conferences as the listener
            self.performSegue(withIdentifier: MainSegueConstant.call,
                              sender: callSettings)
            return
        }
        QBAVCallPermissions.check(with: conferenceType) { granted in
            guard granted == true else { return }
            self.performSegue(withIdentifier: MainSegueConstant.call,
                              sender: callSettings)
        }
    }
}
